File: Design_Patterns/BehaviouralDesignPattern/Stratergy_Design_Pattern/java/with_strategy_pattern/StrategyPattern.java
/**
 * STRATEGY DESIGN PATTERN - Vehicle Example
 * 
 * Problem: Without Strategy Pattern
 * ====================================
 * When using traditional inheritance, if multiple vehicle types need the same
 * special behavior (e.g., SportsCar and OffRoadVehicle both need sports drive),
 * we end up duplicating code in multiple classes.
 * 
 * Solution: Strategy Pattern
 * ==========================
 * - Define a family of algorithms (drive strategies)
 * - Encapsulate each algorithm in a separate class
 * - Make them interchangeable through a common interface
 * - Use composition instead of inheritance
 * 
 * Benefits:
 * - No code duplication
 * - Easy to add new strategies
 * - Runtime flexibility to change behavior
 */

// ============================================================================
// STRATEGY INTERFACE
// ============================================================================

/**
 * Strategy Interface - Defines the contract for all drive strategies
 */
interface DriveStrategy {
    void drive();
}

// ============================================================================
// CONCRETE STRATEGIES
// ============================================================================

/**
 * Normal Drive Strategy - Used by regular vehicles
 */
class NormalDriveStrategy implements DriveStrategy {
    @Override
    public void drive() {
        System.out.println("Normal drive capability");
    }
}

/**
 * Sports Drive Strategy - Used by high-performance vehicles
 */
class SportsDriveStrategy implements DriveStrategy {
    @Override
    public void drive() {
        System.out.println("Sports drive capability - Special driving features!");
    }
}

// ============================================================================
// CONTEXT - Vehicle Base Class
// ============================================================================

/**
 * Vehicle class uses composition to delegate drive behavior to a strategy
 * This is the key principle: "HAS-A" relationship instead of "IS-A"
 */
class Vehicle {
    // Strategy object - composition
    private DriveStrategy driveStrategy;
    
    // Constructor injection - allows setting strategy at creation time
    public Vehicle(DriveStrategy driveStrategy) {
        this.driveStrategy = driveStrategy;
    }
    
    // Delegates to the strategy
    public void drive() {
        driveStrategy.drive();
    }
}

// ============================================================================
// CONCRETE VEHICLES
// ============================================================================

/**
 * Normal Vehicle - Uses normal drive strategy
 */
class NormalVehicle extends Vehicle {
    public NormalVehicle() {
        super(new NormalDriveStrategy());
    }
}

/**
 * Sports Vehicle - Uses sports drive strategy
 */
class SportsVehicle extends Vehicle {
    public SportsVehicle() {
        super(new SportsDriveStrategy());
    }
}

/**
 * OffRoad Vehicle - Also uses sports drive strategy
 * Notice: No code duplication! Both SportsVehicle and OffRoadVehicle
 * share the same SportsDriveStrategy instance behavior
 */
class OffRoadVehicle extends Vehicle {
    public OffRoadVehicle() {
        super(new SportsDriveStrategy());
    }
}

// ============================================================================
// DEMO
// ============================================================================

public class StrategyPattern {
    public static void main(String[] args) {
        System.out.println("=== Strategy Design Pattern Demo ===\n");
        
        // Create different vehicles
        Vehicle normalVehicle = new NormalVehicle();
        Vehicle sportsVehicle = new SportsVehicle();
        Vehicle offRoadVehicle = new OffRoadVehicle();
        
        // Each vehicle uses its assigned strategy
        System.out.print("Normal Vehicle: ");
        normalVehicle.drive();
        
        System.out.print("Sports Vehicle: ");
        sportsVehicle.drive();
        
        System.out.print("OffRoad Vehicle: ");
        offRoadVehicle.drive();
        
        System.out.println("\n✓ Notice: Sports and OffRoad vehicles share the same strategy");
        System.out.println("✓ No code duplication!");
    }
}

File: Design_Patterns/BehaviouralDesignPattern/Stratergy_Design_Pattern/python/with_strategy_pattern/strategy_pattern.py
"""
STRATEGY DESIGN PATTERN - Vehicle Example (Python)

Problem: Without Strategy Pattern
====================================
When using traditional inheritance, if multiple vehicle types need the same
special behavior (e.g., SportsCar and OffRoadVehicle both need sports drive),
we end up duplicating code in multiple classes.

Solution: Strategy Pattern
==========================
- Define a family of algorithms (drive strategies)
- Encapsulate each algorithm in a separate class
- Make them interchangeable through a common interface
- Use composition instead of inheritance

Benefits:
- No code duplication
- Easy to add new strategies
- Runtime flexibility to change behavior
"""

from abc import ABC, abstractmethod

# ============================================================================
# STRATEGY INTERFACE
# ============================================================================

class DriveStrategy(ABC):
    """
    Strategy Interface - Defines the contract for all drive strategies
    """
    
    @abstractmethod
    def drive(self):
        """Abstract method that must be implemented by concrete strategies"""
        pass


# ============================================================================
# CONCRETE STRATEGIES
# ============================================================================

class NormalDriveStrategy(DriveStrategy):
    """
    Normal Drive Strategy - Used by regular vehicles
    """
    
    def drive(self):
        print("Normal drive capability")


class SportsDriveStrategy(DriveStrategy):
    """
    Sports Drive Strategy - Used by high-performance vehicles
    """
    
    def drive(self):
        print("Sports drive capability - Special driving features!")


# ============================================================================
# CONTEXT - Vehicle Base Class
# ============================================================================

class Vehicle:
    """
    Vehicle class uses composition to delegate drive behavior to a strategy
    This is the key principle: "HAS-A" relationship instead of "IS-A"
    """
    
    def __init__(self, drive_strategy: DriveStrategy):
        """
        Constructor injection - allows setting strategy at creation time
        
        Args:
            drive_strategy: The drive strategy to use
        """
        self.drive_strategy = drive_strategy
    
    def drive(self):
        """Delegates to the strategy"""
        self.drive_strategy.drive()


# ============================================================================
# CONCRETE VEHICLES
# ============================================================================

class NormalVehicle(Vehicle):
    """
    Normal Vehicle - Uses normal drive strategy
    """
    
    def __init__(self):
        super().__init__(NormalDriveStrategy())


class SportsVehicle(Vehicle):
    """
    Sports Vehicle - Uses sports drive strategy
    """
    
    def __init__(self):
        super().__init__(SportsDriveStrategy())


class OffRoadVehicle(Vehicle):
    """
    OffRoad Vehicle - Also uses sports drive strategy
    Notice: No code duplication! Both SportsVehicle and OffRoadVehicle
    share the same SportsDriveStrategy behavior
    """
    
    def __init__(self):
        super().__init__(SportsDriveStrategy())


# ============================================================================
# DEMO
# ============================================================================

def main():
    print("=== Strategy Design Pattern Demo ===\n")
    
    # Create different vehicles
    normal_vehicle = NormalVehicle()
    sports_vehicle = SportsVehicle()
    offroad_vehicle = OffRoadVehicle()
    
    # Each vehicle uses its assigned strategy
    print("Normal Vehicle: ", end="")
    normal_vehicle.drive()
    
    print("Sports Vehicle: ", end="")
    sports_vehicle.drive()
    
    print("OffRoad Vehicle: ", end="")
    offroad_vehicle.drive()
    
    print("\n✓ Notice: Sports and OffRoad vehicles share the same strategy")
    print("✓ No code duplication!")


if __name__ == "__main__":
    main()
